RecoVertex/PrimaryVertexProducer/src/AdaptiveChisquarePrimaryVertexFitter.cc

```diff
@@ -0,0 +1,540 @@
+    for (unsigned int i = 0; i < nt; i++) {
+      auto &ti = trackinfo[i];
+      const double zrange = zrange_scale * ti.dzError;
+      if (std::abs(zv[k] - ti.zpca) < z_cutoff_) {
+        const double dztrk = ti.b2 + xv[k] * ti.H2[0] + yv[k] * ti.H2[1] - zv[k];
+        if (std::abs(dztrk) < zrange) {
+          tkmap.emplace_back(i);
+          tkweight.emplace_back(0.);
+        }
+      }
+    }
+  }
+  tkfirstv.emplace_back(tkmap.size());  // extra entry, simplifies loops, every vertex has a "successor" now
+}
+
+void AdaptiveChisquarePrimaryVertexFitter::fill_weights(const reco::BeamSpot &beamspot, double beta) {
+  // multi-vertex version
+  unsigned const int nt = trackinfo.size();
+  unsigned const int nv = xv.size();
+
+  const double Z_cutoff = vdt::fast_exp(-0.5 * beta * chi_cutoff_ * chi_cutoff_);
+  std::vector<double> Z_track(nt, Z_cutoff);
+  const double argmax = 0.5 * beta * chi_cutoff_ * chi_cutoff_ * 5;
+
+  // evaluate and cache track-vertex assignment chi**2 for all clusters and sum up Z
+  for (unsigned int k = 0; k < nv; k++) {
+    for (unsigned int j = tkfirstv[k]; j < tkfirstv[k + 1]; j++) {
+      const unsigned int i = tkmap[j];
+      double arg = 0.5 * beta * track_in_vertex_chsq(trackinfo[i], xv[k], yv[k], zv[k]);
+      if (arg < argmax) {
+        const double e = vdt::fast_exp(-arg);
+        tkweight[j] = e;  // must later be normalized by the proper Z_track[i]
+        Z_track[i] += e;  // sum up exponentials to normalize
+      } else {
+        tkweight[j] = 0.;
+      }
+    }
+  }
+
+  // now we have the partition function, Z_i and can evaluate assignment probabilities (aka weights)
+  for (unsigned int j = 0; j < tkmap.size(); j++) {
+    const unsigned int i = tkmap[j];
+#ifdef PVT_DEBUG
+    assert((i < nt) && "tkmap out of range");
+    assert((tkmap.size() == tkweight.size()) && "map and list not aliged");
+#endif
+    tkweight[j] /= Z_track[i];
+  }
+}
+
+bool AdaptiveChisquarePrimaryVertexFitter::clean() {
+  /* in multi-vertex fitting, nearby vertices can fall on top of each other, 
+     even when the initial seeds don't, some kind of duplicate removal is required
+     the approach in this method is similar to the method applied in clustering:
+     at least two tracks with a weight above a threshold (trkweight_threshold) are required.
+     vertices that don't fulfill this are either insignficant or very close
+     to another vertex
+   */
+  const double trkweight_threshold = 0.7;
+  unsigned int nv = xv.size();
+  if (nv < 2)
+    return false;
+
+  // sum of weights per vertex
+  std::vector<double> wsumhi(nv, 0);
+  for (unsigned int k = 0; k < nv; k++) {
+    for (unsigned int j = tkfirstv[k]; j < tkfirstv[k + 1]; j++) {
+      if (tkweight[j] > trkweight_threshold)
+        wsumhi[k] += tkweight[j];
+    }
+  }
+
+  double dzmin = 0;
+  unsigned int k_dzmin = 0;
+  for (unsigned int k = 0; k < nv - 1; k++) {
+    if ((k == 0) || (std::abs(zv[k + 1] - zv[k]) < dzmin)) {
```

Comment:
Might as well just define `double dz = std::abs(zv[k +1] - zv[k])` rather than evaluate it twice